pkg/cache/sql/informer/listoption_indexer.go

```diff
@@ -285,76 +305,92 @@ func (l *ListOptionIndexer) ListByOptions(ctx context.Context, lo ListOptions, p
 	}
 
 	if len(orderByClauses) > 0 {
-		stmt += "\n  ORDER BY "
-		stmt += strings.Join(orderByClauses, ", ")
+		query += "\n  ORDER BY "
+		query += strings.Join(orderByClauses, ", ")
 	} else {
 		// make sure one default order is always picked
 		if l.namespaced {
-			stmt += "\n  ORDER BY f.\"metadata.namespace\" ASC, f.\"metadata.name\" ASC "
+			query += "\n  ORDER BY f.\"metadata.namespace\" ASC, f.\"metadata.name\" ASC "
 		} else {
-			stmt += "\n  ORDER BY f.\"metadata.name\" ASC "
+			query += "\n  ORDER BY f.\"metadata.name\" ASC "
 		}
 	}
 
-	// 3- Pagination: LIMIT clause (from lo.Pagination and/or lo.ChunkSize/lo.Resume)
+	// 4- Pagination: LIMIT clause (from lo.Pagination and/or lo.ChunkSize/lo.Resume)
+
+	// before proceeding, save a copy of the query and params without LIMIT/OFFSET
+	// for COUNTing all results later
+	countQuery := fmt.Sprintf("SELECT COUNT(*) FROM (%s)", query)
+	countParams := params[:]
+
 	limitClause := ""
-	offsetClause := ""
 	// take the smallest limit between lo.Pagination and lo.ChunkSize
 	limit := lo.Pagination.PageSize
 	if limit == 0 || (lo.ChunkSize > 0 && lo.ChunkSize < limit) {
 		limit = lo.ChunkSize
 	}
 	if limit > 0 {
 		limitClause = "\n  LIMIT ?"
-		// note: retrieve one extra row. If it comes back, then there are more pages and a continueToken should be created
-		params = append(params, limit+1)
+		params = append(params, limit)
 	}
 
 	// OFFSET clause (from lo.Pagination and/or lo.Resume)
+	offsetClause := ""
 	offset := 0
 	if lo.Resume != "" {
 		offsetInt, err := strconv.Atoi(lo.Resume)
 		if err != nil {
-			return nil, "", err
+			return nil, 0, "", err
 		}
 		offset = offsetInt
 	}
 	if lo.Pagination.Page >= 1 {
 		offset += lo.Pagination.PageSize * (lo.Pagination.Page - 1)
 	}
-
 	if offset > 0 {
 		offsetClause = "\n  OFFSET ?"
 		params = append(params, offset)
 	}
 
-	stmt += limitClause
-	stmt += offsetClause
-
-	// log the final query
-	logrus.Debugf("ListOptionIndexer prepared statement: %v", stmt)
+	// assemble and log the final query
+	query += limitClause
+	query += offsetClause
+	logrus.Debugf("ListOptionIndexer prepared statement: %v", query)
 	logrus.Debugf("Params: %v", params...)
 
 	// execute
-	prepStmt := l.Prepare(stmt)
-	defer l.CloseStmt(prepStmt)
-	rows, err := l.QueryForRows(ctx, prepStmt, params...)
+	stmt := l.Prepare(query)
+	defer l.CloseStmt(stmt)
+	rows, err := l.QueryForRows(ctx, stmt, params...)
 	if err != nil {
-		return nil, "", err
+		return nil, 0, "", fmt.Errorf("while executing query: %s got error: %w", query, err)
 	}
 	items, err := l.ReadObjects(rows, l.GetType(), l.GetShouldEncrypt())
 	if err != nil {
-		return nil, "", err
+		return nil, 0, "", err
+	}
+
+	total := len(items)
+	// if limit or offset were set, execute counting of all rows
+	if limit > 0 || offset > 0 {
+		countStmt := l.Prepare(countQuery)
+		defer l.CloseStmt(countStmt)
+		rows, err = l.QueryForRows(ctx, countStmt, countParams...)
```

Comment:
Actually, let me check but I think both those query for rows need to be in a transaction and its not clear to me that they are.

Reply:
I'd still like some input from @moio about this. While we're not writing to the DB, it seems like we should be returning a consistent result. However, it might not be critical if the worse that can happen is the UI receiving a more up-to-date value.. I'm okay with merging this as is to get things moving.

Reply:
I will fix that in a follow-up PR, thanks for spotting